include/internal/quic_stream.h

```diff
@@ -0,0 +1,270 @@
+ * to ossl_quic_sstream_get_stream_frame. The start and end values denote logical byte
+ * numbers and are inclusive.
+ *
+ * If the lost frame had the FIN bit set, you must also call
+ * ossl_quic_sstream_mark_lost_fin() after calling this function.
+ *
+ * Returns 1 on success and 0 on error (e.g. if end < start).
+ */
+int ossl_quic_sstream_mark_lost(QUIC_SSTREAM *qss,
+                                uint64_t start,
+                                uint64_t end);
+
+/*
+ * (RX/ACKM use.) Informs the QSS that a STREAM frame with the FIN bit set was
+ * lost.
+ *
+ * Returns 1 on success and 0 on error.
+ */
+int ossl_quic_sstream_mark_lost_fin(QUIC_SSTREAM *qss);
+
+/*
+ * (RX/ACKM use.) Marks a logical range of the send stream as having been
+ * acknowledged, meaning that the storage for the data in that range of the
+ * stream can be now recycled and neither that logical range of the stream nor
+ * any subset of it can be retransmitted again. The start and end values are
+ * inclusive.
+ *
+ * If the acknowledged frame had the FIN bit set, you must also call
+ * ossl_quic_sstream_mark_acked_fin() after calling this function.
+ *
+ * Returns 1 on success and 0 on error (e.g. if end < start).
+ */
+int ossl_quic_sstream_mark_acked(QUIC_SSTREAM *qss,
+                                 uint64_t start,
+                                 uint64_t end);
+
+/*
+ * (RX/ACKM use.) Informs the QSS that a STREAM frame with the FIN bit set
+ * was acknowledged.
+ *
+ * Returns 1 on success and 0 on error.
+ */
+int ossl_quic_sstream_mark_acked_fin(QUIC_SSTREAM *qss);
+
+/*
+ * (Front end use.) Appends user data to the stream. The data is copied into the
+ * stream. The amount of data consumed from buf is written to *consumed on
+ * success (short writes are possible). The amount of data which can be written
+ * can be determined in advance by calling the ossl_quic_sstream_get_buffer_avail()
+ * function; data is copied into an internal ring buffer of finite size.
+ *
+ * If the buffer is full, this should be materialised as a backpressure
+ * condition by the front end. This is not considered a failure condition;
+ * *consumed is written as 0 and the function returns 1.
+ *
+ * Returns 1 on success or 0 on failure.
+ */
+int ossl_quic_sstream_append(QUIC_SSTREAM *qss,
+                             const unsigned char *buf,
+                             size_t buf_len,
+                             size_t *consumed);
+
+/*
+ * Marks a stream as finished. ossl_quic_sstream_append() may not be called anymore
+ * after calling this.
+ */
+void ossl_quic_sstream_fin(QUIC_SSTREAM *qss);
+
+/*
+ * Resizes the internal ring buffer. All stream data is preserved safely.
+ *
+ * This can be used to expand or contract the ring buffer, but not to contract
+ * the ring buffer below the amount of stream data currently stored in it.
+ * Returns 1 on success and 0 on failure.
+ *
+ * IMPORTANT: Any buffers referenced by iovecs output by
```

Comment:
Would a double ended queue be worthwhile considering instead of a ring buffer?  It would avoid this caveat.  It might also avoid issues wrapping around the end of the buffer.

Another option might be a mem BIO.

Reply:
SSTREAM is merged now, but we can consider the merits of changing the underlying data structure later. I don't anticipate this pointer validation causing any issues for now since we will only be handling the output of `get_stream_frame` in the midst of specific operations like TX packetisation, etc.. We should not need any long-lived references to returned data.